src/circuit/integer/assert_not_zero.rs

```diff
@@ -0,0 +1,91 @@
+use super::IntegerChip;
+use crate::circuit::main_gate::{CombinationOption, MainGateInstructions, Term};
+use crate::circuit::{AssignedInteger, AssignedValue};
+use crate::BIT_LEN_LIMB;
+use halo2::arithmetic::FieldExt;
+use halo2::circuit::Region;
+use halo2::plonk::Error;
+use num_bigint::BigUint as big_uint;
+use std::convert::TryInto;
+
+impl<W: FieldExt, N: FieldExt> IntegerChip<W, N> {
+    pub(crate) fn _assert_not_zero(&self,
+        region: &mut Region<'_, N>,
+        a: &AssignedInteger<N>,
+        offset: &mut usize
+    ) -> Result<(), Error> {
+        let main_gate = self.main_gate();
+        let (zero, one) = (N::zero(), N::one());
+
+        // reduce result (r) should be less than 2 * wrong_modulus,
+        // so we only need to assert r <> 0 and r <> wrong_modulus.
+        let r = self._reduce(region, a, offset)?;
+
+        // Sanity check.
+        // This algorithm requires that wrong modulus * 2 <= native modulus * 2 ^ BIT_LEN_LIMB.
+        let two_pow_limb_bits_minus_1 = big_uint::from(2u64).pow((BIT_LEN_LIMB - 1).try_into().unwrap());
+        assert!(self.rns.wrong_modulus.clone() <= self.rns.native_modulus.clone() * two_pow_limb_bits_minus_1);
+
+        // r = 0 <-> r % 2 ^ 64 = 0 /\ r % native_modulus = 0
+        // r <> 0 <-> r % 2 ^ 64 <> 0 \/ r % native_modulus <> 0
+        // r <> 0 <-> invert(r.limb(0)) \/ invert(r.native())
+        let (_, cond_zero_0) = main_gate.invert(region, r.limb(0), offset)?;
```

Comment:
Can we use main_gate.is_zero here to make it more expressive

Reply:
Done